File: bacnet-sc-reference-stack-code/dev/src/java/com/example/demo/DemoCommands.java
// Copyright, distribution, and usage are defined by the top level LICENSE file.
package com.example.demo;

import dev.bscs.bacnet.applications.BACnetCommands;
import dev.bscs.bacnet.stack.*;
import dev.bscs.bacnet.stack.data.BACnetPropertyIdentifier;
import dev.bscs.bacnet.stack.data.base.BACnetData;
import dev.bscs.bacnet.stack.data.base.BACnetObjectIdentifier;
import dev.bscs.bacnet.stack.services.ReadPropertyClient;
import dev.bscs.bacnet.stack.services.WhoIsClient;
import dev.bscs.common.*;
import dev.bscs.events.EventListener;
import dev.bscs.events.EventLoop;
import dev.bscs.events.EventType;

import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

/**
 * This is a command that was made just for a demonstration of how to create a command.
 * It is of limited usefulness but is included here because it was referenced in a webinar.
 * It can be included without code changes in any application with a line in the configuration like this:
 *     common.commandClass = com.example.demo.DemoCommands
 * @author drobin
 */

public class DemoCommands implements CommandProcessor.CommandSet {

    public void addCommands() { CommandProcessor.addCommand(new WixCommand()); }

    private static class WixCommand extends Command implements EventListener {
        public WixCommand() { super("wix","wix","WhoIsExtended - a WhoIs followed by a ReadProperty");}
        Queue<Binding> queue = new ConcurrentLinkedQueue<>();
        Timer timer = new Timer();
        @Override public void handleEvent(Object source, EventType eventType, Object... args) {
            Binding binding;
            if (timer.remaining() == 0 && (binding = queue.poll()) != null) {
                timer.start(2000L);
                Shell.print(Formatting.toNetMac(binding.dnet,binding.dadr));
                readPropertyClient.request(
                        BACnetCommands.getSelectedDevice(), binding.dnet, binding.dadr,
                        BACnetObjectIdentifier.combine(8,0x3FFFFF),
                        BACnetPropertyIdentifier.OBJECT_NAME, -1
                );
            }
        }
        ReadPropertyClient readPropertyClient = new ReadPropertyClient() {
            @Override protected void success(Device device, BACnetData value, AuthData auth) {
                Shell.println(" name="+value.toString()+" auth="+auth);
                timer.clear();
                EventLoop.emit(this, WixCommand.this, EventLoop.EVENT_MAINTENANCE);
            }
            @Override protected void failure(Device device, Failure failure, AuthData auth) {
                Shell.println(" failure="+failure);
                timer.clear();
                EventLoop.emit(this, WixCommand.this, EventLoop.EVENT_MAINTENANCE);
            }
        };
        WhoIsClient whoIsClient = new WhoIsClient() {
            @Override protected void success(Device device, Binding binding, AuthData auth) {
                queue.offer(binding);
                EventLoop.emit(this, WixCommand.this, EventLoop.EVENT_MAINTENANCE);
            }
        };
        @Override public boolean execute(String[] words) {
            // We want to execute a global Who-Is,
            // and then for each response, execute a ReadProperty for the Device object's name.
            whoIsClient.request(BACnetCommands.getSelectedDevice(),65535,null);
            EventLoop.addMaintenance(this);
            return false; // returning false indicates that this command is not "done"
        }
        @Override public void cancel() {
            EventLoop.removeMaintenance(this);
            whoIsClient.cancel(BACnetCommands.getSelectedDevice());
        }
    }

}

File: bacnet-sc-reference-stack-code/scripts/DemoCommands.py
from dev.bscs.common import Shell
from dev.bscs.common import Timer
from dev.bscs.common import Command
from dev.bscs.common import CommandProcessor
from dev.bscs.common import Formatting
from dev.bscs.bacnet.applications import BACnetCommands
from dev.bscs.bacnet.stack.data import BACnetPropertyIdentifier
from dev.bscs.bacnet.stack.data.base import BACnetObjectIdentifier
from dev.bscs.bacnet.stack.services import ReadPropertyClient
from dev.bscs.bacnet.stack.services import WhoIsClient
from dev.bscs.events import EventListener
from dev.bscs.events import EventLoop
from java.lang import String
from java.util.concurrent import ConcurrentLinkedQueue
from jarray import array

# This is just a demonstration of how to add commands from Python.
# If you run
#      $ py load DempCommands.py
# the static code here will add two commands to the terminal.
# You can see the available commands with
#      $ list
# You can see the dynamic nature of this scripy be changing this file and just running the "load" command again.
# The changed commands become immediately available without needing to restart.

class PwixCommand(Command):
    
    class MyListener(EventListener):
        def __init__(self,parent):
            self.parent = parent
        def handleEvent(self, source, eventType, *args):
            if self.parent.timer.remaining() == 0:
                binding = self.parent.queue.poll()
                if binding != None:
                    self.parent.timer.start(2000)
                    Shell.print(Formatting.toNetMac(binding.dnet,binding.dadr))
                    self.parent.readPropertyClient.request(BACnetCommands.getSelectedDevice(), binding.dnet, binding.dadr,BACnetObjectIdentifier.combine(8,0x3FFFFF),BACnetPropertyIdentifier.OBJECT_NAME, -1)
    
    class MyWhoIsClient(WhoIsClient):
        def __init__(self,parent):
            self.parent = parent
        def success(self, device, binding, auth):
            self.parent.queue.offer(binding)
            EventLoop.emit(self, self.parent.listener, EventLoop.EVENT_MAINTENANCE)

            
    class MyReadPropertyClient(ReadPropertyClient):
        def __init__(self,parent):
            self.parent = parent
        def success(self, device, value, auth):
            Shell.println(" name=" + value.toString() + " auth=" + auth.toString())
            self.parent.timer.clear()
            EventLoop.emit(self, self.parent.listener, EventLoop.EVENT_MAINTENANCE)
        def failure(self, device, failure, auth):
            Shell.println(" failure="+failure)
            self.parent.timer.clear()

    def __init__(self):
        Command.__init__(self,"pwix","pwix",array(["WhoIsExtended - a WhoIs followed by a ReadProperty"],String))
        self.queue = ConcurrentLinkedQueue()
        self.timer = Timer()
        self.readPropertyClient = PwixCommand.MyReadPropertyClient(self)
        self.listener = PwixCommand.MyListener(self)
        self.whoIsClient = PwixCommand.MyWhoIsClient(self)

    def execute(self, words):
        self.whoIsClient.request(BACnetCommands.getSelectedDevice(),65535,None)
        EventLoop.addMaintenance(self.listener)
        return False # keeps prompt open for long running operation

    def cancel(self):
        self.whoIsClient.cancel(BACnetCommands.getSelectedDevice())
        EventLoop.removeMaintenance(self.listener)
        
        
class HiCommand(Command):
    def __init__(self):
        Command.__init__(self,"hi","hi",array(["hi - say Hi!"],String))

    def execute(self, words):
        Shell.println("Hi!")
        return True # command is finished, re-prompt

# if you reload this script, the commands gets properly replaced with new versions.
CommandProcessor.addCommand(PwixCommand()) 
CommandProcessor.addCommand(HiCommand()) 

  
    
           



